components/renku_data_services/notebooks/api/classes/k8s_client.py

```diff
@@ -2,240 +2,215 @@
 
 import base64
 import json
-from typing import Any, Optional, cast
+import logging
+from contextlib import suppress
+from typing import Any, Generic, Optional, TypeVar, cast
 from urllib.parse import urljoin
 
-import requests
-from kubernetes import client
-from kubernetes.client.exceptions import ApiException
-from kubernetes.client.models import V1Container, V1DeleteOptions
-from kubernetes.config import load_config
-from kubernetes.config.config_exception import ConfigException
-from kubernetes.config.incluster_config import SERVICE_CERT_FILENAME, SERVICE_TOKEN_FILENAME, InClusterConfigLoader
-from sanic.log import logger
+import httpx
+from kr8s import NotFoundError, ServerError
+from kr8s.asyncio.objects import APIObject, Pod, Secret, StatefulSet
+from kubernetes.client import ApiClient, V1Container, V1Secret
 
-from ...errors.intermittent import (
+from renku_data_services.errors import errors
+from renku_data_services.notebooks.api.classes.auth import GitlabToken, RenkuTokens
+from renku_data_services.notebooks.crs import AmaltheaSessionV1Alpha1, JupyterServerV1Alpha1
+from renku_data_services.notebooks.errors.intermittent import (
     CannotStartServerError,
     DeleteServerError,
     IntermittentError,
     JSCacheError,
     PatchServerError,
 )
-from ...errors.programming import ProgrammingError
-from ...errors.user import MissingResourceError
-from ...util.kubernetes_ import find_env_var
-from ...util.retries import retry_with_exponential_backoff
-from .auth import GitlabToken, RenkuTokens
+from renku_data_services.notebooks.errors.programming import ProgrammingError
+from renku_data_services.notebooks.errors.user import MissingResourceError
+from renku_data_services.notebooks.util.kubernetes_ import find_env_var
+from renku_data_services.notebooks.util.retries import (
+    retry_with_exponential_backoff_async,
+)
+
+sanitize_for_serialization = ApiClient().sanitize_for_serialization
+
+
+# NOTE The type ignore below is because the kr8s library has no type stubs, they claim pyright better handles type hints
+class AmaltheaSessionV1Alpha1Kr8s(APIObject):  # type: ignore
```

Comment:
question: I'm a bit confused why this uses `Kr8s` instead of the more common `k8s`? even the doc string uses k8s. that applies to this whole file

Reply:
I reworded the docstring. Normal python k8s has zero async support - the main reason for using kr8s is that they support async.

Reply:
Or are you talking about renaming the title of the class to be `AmaltheaSessionV1Alpha1K8s` instead of `AmaltheaSessionV1Alpha1Kr8s` i.e. just removing the `r` from the name?

Reply:
yes the title as well. it's really minor, just thought we should stick to one naming convention